### PR TITLE
Update the node selector implementation to handle ordering of select calls

### PR DESCRIPTION
**Pull request by vietj**
Motivation:
    
The `DefaultNodeSelector` implementation relies on a promise to broadcast the resolution of an address to its waiters. Ordering is therefore implemented as a per Vert.x context construct prior selector resolution, ordering is achieved per vertx context and per address using a temporary queue.
    
Ordering is required because the node selector and the underlying cluster manager use a cached promise to resolve the multiple select calls which does not maintain ordering. Since we do control the node selector implementation, ordering can be achieved in the node selector itself, providing the ordering required that the cluster manager does not provide.
    
Changes:
    
Updated the node selector implementation to implement ordering.
    
A linked list of select result promises is kept in a concurrent map until the cluster manager response is obtained, when the promise resolves the result promise list is resolved in arrival order an the list of registration is cached with the round robin selector for new node selector requests.
    
As consequences ordering is implemented based on the order of node selector calls.

### CODE REVIEW

**Comment by vietj** [comment]
@tsegismont can you have a look ?

**Review by tsegismont — COMMENTED** [review]
`Serializer`  was in `ClusteredEventBus` because we wanted ordering to happen in all cases, even when users provided their own implementation of `NodeSelector`.

Now it seems it's still possible, by extending `DefaultNodeSelector2` (can you think of a better name please).

Also, providing a custom node selector is no longer possible without using impl package classes, so I assume none of this works in JPMS land.

I think we should either:
- stop supporting custom node selectors
- make it work properly (using internal package classes that are accessible, even in JPMS land)

No strong opinion from my side. In any case there are breaking changes that need to be documented.

File: vertx-core/src/test/java/io/vertx/tests/eventbus/ClusteredEventBusTest.java [hunk 1]
```diff
@@ -671,4 +673,27 @@ public void testMultiHeaders() {
     await();
 
   }
+
+  @Test
+  public void testPreserveMessageOrderingOnContext() {
```
[the remark this answers is not in the document]
Comment (vietj) [reply vertx-core/src/test/java/io/vertx/tests/eventbus/ClusteredEventBusTest.java:678]: I didn't find it, perhaps in cluster manager implems?

**Comment by vietj** [comment]
I think we should move node selector more internally but week them in case we want to resurect them.

**Comment by vietj** [comment]
NOTE: this PR needs also to be tested against cluster manager implementations

File: vertx-core/src/test/java/io/vertx/test/fakecluster/FakeClusterManager.java [hunk 2]
```diff
@@ -49,12 +50,21 @@ public class FakeClusterManager implements ClusterManager {
   private NodeListener nodeListener;
   private RegistrationListener registrationListener;
   private VertxInternal vertx;
+  private volatile long getRegistrationsLatency = 0L;
```
[the remark this answers is not in the document]
Comment (vietj) [reply vertx-core/src/test/java/io/vertx/test/fakecluster/FakeClusterManager.java:53]: the idea is that it is called after the "getRegistrations" method because other operations might use a delay in future for testing purpose

File: vertx-core/src/test/java/io/vertx/test/fakecluster/FakeClusterManager.java [hunk 3]
```diff
@@ -49,12 +50,21 @@ public class FakeClusterManager implements ClusterManager {
...
     this.vertx = (VertxInternal) vertx;
   }
 
+  public long getRegistrationsLatency() {
+    return getRegistrationsLatency;
+  }
+
+  public void getRegistrationsLatency(long getRegistrationsLatency) {
+    this.getRegistrationsLatency = getRegistrationsLatency;
+  }
```
[the remark this answers is not in the document]
Comment (vietj) [reply vertx-core/src/test/java/io/vertx/test/fakecluster/FakeClusterManager.java:66]: the idea is that it is called after the "getRegistrations" method because other operations might use a delay in future for testing purpose

File: vertx-core/src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java [hunk 4]
```diff
@@ -189,25 +189,20 @@ protected <T> void sendOrPub(ContextInternal ctx, MessageImpl<?, T> message, Del
     } else if (options.isLocalOnly()) {
       sendLocally(message, writePromise);
     } else {
-      Serializer serializer = Serializer.get(ctx);
```
[the remark this answers is not in the document]
Comment (vietj) [reply vertx-core/src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java:192]: yes will do that

File: vertx-core/src/test/java/io/vertx/tests/eventbus/ClusteredEventBusTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vietj) [reply vertx-core/src/test/java/io/vertx/tests/eventbus/ClusteredEventBusTest.java:674]: I think we do have with `MessageQueueOnWorkerThreadTest` which test in the case of worker when we had the issue. FWIW in my early implementation I actually reversed the order in which message are replied after cumulation in the new node selector implementation and the test did not fail.

I think the new test is better as it is because it introduces latency at the cluster manager level which is is what is likely to happen with message re-ordering.

**Comment by tsegismont** [comment]
> node selector is moved as implementation but remains a thing because it is useful to test a couple of edge cases like `WriteHandlerLookupFailureTest`.

Please report this to the breaking changes wiki
